libtransact/src/context/execution_context.rs

```diff
@@ -0,0 +1,199 @@
+use crate::receipts::Event;
+use std::fmt::Debug;
+use uuid::Uuid;
+
+use crate::receipts::StateChange;
+
+#[derive(Debug, Clone, Default)]
+pub struct Context<K, V> {
+    base_contexts: Vec<ContextId>,
+    state_changes: Vec<StateChange<K, V>>,
+    id: ContextId,
+    data: Vec<Vec<u8>>,
+    events: Vec<Event>,
+    state_id: String,
+}
+
+impl<K: PartialEq + Clone + Debug, V: Clone + Debug> Context<K, V> {
+    pub fn new(state_id: &str, base_contexts: Vec<ContextId>) -> Self {
+        Context {
+            base_contexts,
+            state_id: state_id.to_string(),
+            state_changes: Vec::new(),
+            id: *Uuid::new_v4().as_bytes(),
+            data: Vec::new(),
+            events: Vec::new(),
+        }
+    }
+
+    pub fn base_contexts(&self) -> &[ContextId] {
+        &self.base_contexts
+    }
+    pub fn events(&self) -> &Vec<Event> {
+        &self.events
+    }
+
+    pub fn state_changes(&self) -> &Vec<StateChange<K, V>> {
+        &self.state_changes
+    }
+
+    pub fn id(&self) -> &ContextId {
+        &self.id
+    }
+
+    pub fn data(&self) -> &Vec<Vec<u8>> {
+        &self.data
+    }
+
+    pub fn state_id(&self) -> &String {
+        &self.state_id
+    }
+
+    pub fn add_event(&mut self, event: Event) {
+        if !self.events().contains(&event) {
+            self.events.push(event);
+        }
+    }
+
+    pub fn add_data(&mut self, data: Vec<u8>) {
+        if !self.data().contains(&data) {
+            self.data.push(data);
+        }
+    }
+
+    /// Adds StateChange::Set without deleting previous StateChanges associated with the Key
+    pub fn set_state(&mut self, key: K, value: V) {
+        let new_state_change = StateChange::Set {
+            key: key.clone(),
+            value: value.clone(),
+        };
+        self.state_changes.push(new_state_change);
+    }
+
+    /// Adds StateChange::Delete and returns the value associated to the key being deleted
+    pub fn delete_state(&mut self, key: K) -> (K, Option<V>) {
+        let mut deleted_value: Option<V> = None;
+        let found_state_change = self
+            .state_changes()
+            .iter()
+            .rev()
+            .cloned()
+            .find(|state_change| state_change.has_key(&key));
+        if let Some(StateChange::Set { value: v, .. }) = found_state_change {
+            deleted_value = Some(v);
+        }
+        let new_state_change: StateChange<_, _> = StateChange::Delete { key: key.clone() };
+        self.state_changes.push(new_state_change);
+        (key.clone(), deleted_value)
+    }
+
+    /// Checks to see if the Key is referenced by any StateChanges within the Context
+    pub fn contains(&self, key: &K) -> bool {
+        let mut exists = false;
+        for state_change in self.state_changes().iter() {
+            match state_change {
+                StateChange::Set { key: k, .. } => {
+                    if k == key {
+                        exists = true;
```

Comment:
You can probably return true here, return false in the next location, and false at the end of the function. (I.e. short-circuit the function)